pybamm/expression_tree/binary_operators.py

```diff
@@ -398,19 +401,33 @@ def __init__(self, left, right):
     def diff(self, variable):
         """ See :meth:`pybamm.Symbol.diff()`. """
         # We shouldn't need this
-        raise NotImplementedError
+        raise NotImplementedError(
+            "diff not implemented for symbol of type 'MatrixMultiplication'"
+        )
 
     def jac(self, variable):
         """ See :meth:`pybamm.Symbol.jac()`. """
-        # I think we only need the case where left is an array and right
+        # We only need the case where left is an array and right
         # is a (slice of a) state vector, e.g. for discretised spatial
-        # operators of the form D @ u
+        # operators of the form D @ u (also catch cases of (-D) @ u)
         left, right = self.orphans
-        if isinstance(left, pybamm.Array):
+        if isinstance(left, pybamm.Array) or (
+            isinstance(left, pybamm.Negate) and isinstance(left.child, pybamm.Array)
+        ):
             left = pybamm.Matrix(csr_matrix(left.evaluate()))
-            return left @ right.jac(variable)
+            jac = left @ right.jac(variable)
+            # Remove domain as matrix multiplies which come from an integral
+            # end up getting the child domains back which throws an error
+            jac.domain = []
```

Comment:
ah nice, I was getting loads of errors like this but didn't think of doing this :)